core/src/main/scala/kafka/server/ClientQuotaManager.scala

```diff
@@ -54,11 +59,84 @@ object ClientQuotaManagerConfig {
   val DefaultQuotaWindowSizeSeconds = 1
   // Purge sensors after 1 hour of inactivity
   val InactiveSensorExpirationTimeSeconds  = 3600
+
+  val UnlimitedQuota = Quota.upperBound(Long.MaxValue)
+  val DefaultClientIdQuotaId = QuotaId.quotaId(None, Some(ConfigEntityName.Default))
+  val DefaultUserQuotaId = QuotaId.quotaId(Some(ConfigEntityName.Default), None)
+  val DefaultUserClientIdQuotaId = QuotaId.quotaId(Some(ConfigEntityName.Default), Some(ConfigEntityName.Default))
+}
+
+object QuotaTypes {
+  val NoQuotas = 0
+  val ClientIdQuotaEnabled = 1
+  val UserQuotaEnabled = 2
+  val UserClientIdQuotaEnabled = 4
+}
+
+object QuotaId {
+
+  val Separator = ':'
+
+  /**
+   * Sanitizes user principal to a safe value for use in MetricName
+   * and as Zookeeper node name
+   */
+  def sanitize(user: String) = {
+    URLEncoder.encode(user, StandardCharsets.UTF_8.name)
+              .replace("*", "%2A") // Metric ObjectName treats * as pattern
+              .replace("+", "%20") // Space URL-encoded as +, replace with percent encoding
+  }
+
+  /**
+   * Decodes sanitized user principal
+   */
+  def desanitize(sanitizedUser: String) = {
+    URLDecoder.decode(sanitizedUser, StandardCharsets.UTF_8.name)
```

Comment:
URL decoder decodes all % encoded sequences, so special treatment is required for * and +. Have added a unit test to verify this too.